### ISSUE TITLE
Use inspect.signature()

### ISSUE THREAD

**Issue by robertvazan** [google/python-fire#159]
Use [inspect.signature()](https://docs.python.org/3.5/library/inspect.html#inspect.signature) if available, because `inspect.getfullargspec()` ignores annotations added by [functools.update_wrapper](https://docs.python.org/3/library/functools.html#functools.update_wrapper) to wrapper functions. Consequently any wrapper applied to functions in the module kills `--help` output and hurts interpretation of function arguments by making all parameters positional.

**Comment by vbarbaresi** [comment]
I had a shot at this: https://github.com/vbarbaresi/python-fire/commit/38c154d9db2c7d2349d1eed8842ec6b93be42060


`inspect.getfullargspec()` is already kind of using `inspect.signature()` behind, just with legacy parameters:
(https://github.com/python/cpython/blob/3.8/Lib/inspect.py#L1090)
`skip_bound_arg=False` and `follow_wrapper_chains=False`

I overrode `inspect.getfullargspec()` with a method that reverses these parameters, and just removed the manual skipping of bound args:

```diff
-  if skip_arg and args:
-    args.pop(0)  # Remove 'self' or 'cls' from the list of arguments.
```
And it solves the `@lru_cache()` decorator case (#145)
But it doesn't solve #190, it looks like `-h` as argument without `--` is another code path

So I'm not really satisfied with it, but I can turn in into a PR if you think it's acceptable.


I should add that the lru_cache() decorator case is rather weird, I couldn't find a minimal reproduction case:

```python
import fire
from functools import update_wrapper


def decorator(maxsize=128, typed=False):
    def decorating_function(user_function):
        wrapper = _wrapper(user_function, maxsize, typed)
        return update_wrapper(wrapper, user_function)
    return decorating_function


def _wrapper(user_function, maxsize, typed):
    def wrapper(*args, **kwargs):
        result = user_function(*args, **kwargs)
        return result
    return wrapper

@decorator()
def foo(name):
    return('Hello %s' % name)

if __name__ == '__main__':
    fire.Fire()
```
This is as close as I could do from the actual lru_cache() decorator, but it doesn't trigger the bug.
So I had to really use functools.lru_cache in the tests..

**Comment by dbieber** [comment]
Resolved by #203 (though looks likely to remain broken in Python 3.4)
